### ISSUE TITLE
Issues with @Transform Execution Due to Upgrade in @nestjs/mapped-types

### ISSUE THREAD

**Issue by marlon-chagas** [nestjs/mapped-types#1517]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Current behavior

When using `IntersectionType` from `@nestjs/swagger` in classes that share properties, the `@Transform` decorator is being executed multiple times for common properties starting from version **2.0.4**. Prior to this version, `@Transform` was executed only once per property. This repeated execution of `@Transform` can lead to unexpected results and inefficiencies, especially when the transformations are complex.

### Minimum reproduction code

https://github.com/marlon-chagas/nestjs-mapped-types-bug

### Steps to reproduce

1. `npm i`
2. `npm run test`

### Expected behavior

I expect the `@Transform` decorator to be executed only once per property, regardless of how many times the property appears in the intersection type.
In the demo repository, on the `rollback-version` [branch](https://github.com/marlon-chagas/nestjs-mapped-types-bug/tree/rollback-version), the version of `@nestjs/mapped-types` was rolled back to `2.0.3`. In this version, the `@Transform` decorator is executed only once per property, regardless of how many times the property is repeated.

### Package version

2.0.4

### Node.js version

20.18.0

### In which operating systems have you tested?

- [ ] macOS
- [ ] Windows
- [X] Linux

### Other

**Note:** In the update of the `@nestjs/mapped-types` library from version 2.0.2 to 2.0.3, a bug was introduced. After upgrading the project, multiple error messages appear where we use `IntersectionType`. This issue may have arisen while attempting to fix another bug that was introduced in version 2.0.3 [(changes)](https://github.com/nestjs/mapped-types/compare/2.0.3...2.0.4#diff-06210431f00fc127c77880b445efe03246cc6c6a8cb9f6c0d95ffbda41d08d57), and nothing was changed in version 2.0.5 to address it.

## `npm run test` in 2.0.3 release
![image](https://github.com/user-attachments/assets/d3ddf964-2f8a-423c-98a5-6a417c464635)

## `npm run test` in ^2.0.4 release
![image](https://github.com/user-attachments/assets/f782ec15-3269-4d39-923d-72898833f241)

**Comment by kevin-riquena** [comment]
+1

**Comment by kamilmysliwiec** [comment]
Would you like to create a PR for this issue?

v2.0.3 change https://github.com/nestjs/mapped-types/commit/c65d434c105e30f2fd35159fd7ecbb497f67b194
v2.0.4 change https://github.com/nestjs/mapped-types/commit/a94aac8b79b8f44280e8c30b1c0d3392fd15bfba

**Comment by marlon-chagas** [comment]
@kamilmysliwiec 

I'd be interested in making the fix and opening a PR, but I'm struggling to understand the reasoning behind the changes made between versions 2.0.2 and 2.0.3. It seems that reverting to 2.0.2 could potentially resolve the issue, but it would also undo the changes introduced in 2.0.3, creating a new breaking change for those updates.

The issue appears to stem from changing this:

```js
metadataMap.set(
  targetClass,
  new Map([...existingRules, ...targetMetadataEntries]),
);
```
to this:
```js
const mergeMap = new Map<string, any[]>();
[existingRules, targetMetadataEntries].forEach((entries) => {
  for (const [valueKey, value] of entries) {
    if (mergeMap.has(valueKey)) {
      if (Array.isArray(mergeMap.get(valueKey))) {
        mergeMap
          .get(valueKey)!
          .push(...(Array.isArray(value) ? value : [value]));
      }
    } else {
      mergeMap.set(valueKey, value);
    }
  }
});
metadataMap.set(targetClass, mergeMap);
```

**Comment by riquena969** [comment]
Hi @kamilmysliwiec,

I’ve come up with a potential solution and can open a PR. However, I have a question about the priority order in `@Transform`. This issue occurs when using `IntersectionType` to extend two classes that have overlapping properties, both of which use `@Transform`. In this case, `@Transform` is applied twice. If the transformations differ between classes (e.g., one transforms "buy" to "BUY" and the other transforms "buy" to "BID"), which transformation should take priority?

Thanks for the guidance!

**Comment by kamilmysliwiec** [comment]
@riquena969 I'd say it doesn’t really matter—if someone’s using `IntersectionType` to merge two classes that have the same set of properties with `@Transform` annotations, that’s already a code smell. In that case, a random transformation order seems justified

**Comment by riquena969** [comment]
@kamilmysliwiec , I opened a PR that prevents decorators from being stacked anymore. This was intentionally introduced for `PickType` in PR https://github.com/nestjs/mapped-types/pull/1213, and I had to add an extra if statement to work around the new behavior that had been inserted so that stacking only does not occur in the case of `IntersectionType`.

If you think it makes sense to merge this change, the PR is https://github.com/nestjs/mapped-types/pull/1523.
